File: src/gpio.c
#include <errno.h>
#include <fcntl.h>
#include <linux/gpio.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

#include "gpio.h"
#include "spi.h"

static char gpio_consumer_name[] = "blimp-ip-over-lora";

void gpio_init(gpio_ctl_t* self,
               char* dev_path,
               uint8_t out_lines_count,
               uint32_t* out_lines,
               uint8_t in_lines_count,
               uint32_t* in_lines) {
  self->fd = open(dev_path, 0);
  if (self->fd < 0) {
    fprintf(stderr, "couldn't open GPIO device %s\nerrno = %d", dev_path,
            errno);
    exit(1);
  }

  if (out_lines_count > 0) {
    struct gpio_v2_line_request out_req;
    memset(&out_req, 0, sizeof(out_req));
    out_req.config.flags = GPIO_V2_LINE_FLAG_OUTPUT;
    out_req.config.num_attrs = 0;  // do we need attributes?
    memcpy(out_req.offsets, out_lines, out_lines_count);
    out_req.num_lines = out_lines_count;
    strncpy(out_req.consumer, gpio_consumer_name, GPIO_MAX_NAME_SIZE);

    IOCTL_WITH_ERR_HANDLING(self->fd, GPIO_V2_GET_LINE_IOCTL, &out_req,
                            "couldn't open GPIO device %s output lines\n",
                            dev_path);
    self->out_lines_fd = out_req.fd;
    if (self->out_lines_fd < 0) {
      fprintf(stderr, "couldn't open GPIO device %s output lines\nerrno = %d",
              dev_path, errno);
      exit(1);
    }
  } else {
    self->out_lines_fd = -1;
  }

  if (in_lines_count > 0) {
    struct gpio_v2_line_request in_req;
    memset(&in_req, 0, sizeof(in_req));
    in_req.config.flags = GPIO_V2_LINE_FLAG_INPUT;
    in_req.config.num_attrs = 0;  // do we need attributes?
    memcpy(in_req.offsets, in_lines, in_lines_count);
    in_req.num_lines = in_lines_count;
    strncpy(in_req.consumer, gpio_consumer_name, GPIO_MAX_NAME_SIZE);

    IOCTL_WITH_ERR_HANDLING(self->fd, GPIO_V2_GET_LINE_IOCTL, &in_req,
                            "couldn't open GPIO device %s input lines\n",
                            dev_path);
    self->in_lines_fd = in_req.fd;
    if (self->in_lines_fd < 0) {
      fprintf(stderr, "couldn't open GPIO device %s in lines\nerrno = %d",
              dev_path, errno);
      exit(1);
    }
  } else {
    self->in_lines_fd = -1;
  }
}

void gpio_deinit(gpio_ctl_t* self) {
  if (self->out_lines_fd >= 0) {
    close(self->out_lines_fd);
  }
  if (self->in_lines_fd >= 0) {
    close(self->in_lines_fd);
  }
  close(self->fd);
}

void gpio_set(gpio_ctl_t* self, uint64_t val, uint64_t mask) {
  struct gpio_v2_line_values vals;
  vals.bits = val;
  vals.mask = mask;
  IOCTL_WITH_ERR_HANDLING(self->out_lines_fd, GPIO_V2_LINE_SET_VALUES_IOCTL,
                          &vals, "couldn't set GPIO values\n");
}

uint64_t gpio_get(gpio_ctl_t* self, uint64_t mask) {
  struct gpio_v2_line_values vals;
  vals.mask = mask;
  IOCTL_WITH_ERR_HANDLING(self->in_lines_fd, GPIO_V2_LINE_GET_VALUES_IOCTL,
                          &vals, "couldn't get GPIO values\n");
  return vals.bits;
}

File: src/gpio.zig
const std = @import("std");
const c = @import("c");
const misc = @import("misc.zig");

const gpio_consumer_name: []const u8 = "blimp-ip-over-lora";

pub const GpioCtrl = struct {
    dev_file: std.Io.File,
    out_lines_buf: [64]u32,
    out_lines: []u32,
    out_lines_fd: i32,
    in_lines_buf: [64]u32,
    in_lines: []u32,
    in_lines_fd: i32,

    pub fn init(io: std.Io, dev_path: []const u8, out_lines: []const u32, in_lines: []const u32) !GpioCtrl {
        if (out_lines.len > 64 or in_lines.len > 64) {
            return error.LengthTooLarge;
        }

        const dev_file = try std.Io.Dir.openFileAbsolute(io, dev_path, .{ .mode = .read_write });
        const fd = dev_file.handle;

        var self = GpioCtrl{
            .dev_file = dev_file,
            .out_lines_buf = std.mem.zeroes([64]u32),
            .out_lines = undefined,
            .out_lines_fd = undefined,
            .in_lines_buf = std.mem.zeroes([64]u32),
            .in_lines = undefined,
            .in_lines_fd = undefined,
        };
        @memcpy(self.out_lines[0..out_lines.len], out_lines);
        @memcpy(self.in_lines[0..in_lines.len], in_lines);

        if (out_lines.len > 0) {
            var out_req = std.mem.zeroes(c.gpio_v2_line_request);
            out_req.config.flags = c.GPIO_V2_LINE_FLAG_OUTPUT;
            out_req.config.num_attrs = 0; // do we need attributes?
            @memcpy(out_req.offsets[0..out_lines.len], out_lines);
            out_req.num_lines = @intCast(out_lines.len);
            @memcpy(out_req.consumer[0..gpio_consumer_name.len], gpio_consumer_name);

            _ = try misc.ioctl_checked(
                fd,
                c.GPIO_V2_GET_LINE_IOCTL,
                @intFromPtr(&out_req),
                "couldn't open GPIO output",
            );
            if (out_req.fd < 0) {
                std.log.err(
                    "Failed to open gpio device {s} output lines with errno = {d}",
                    .{ dev_path, std.c._errno().* },
                );
                return error.GpioLineOpenFailed;
            }
            self.out_lines_fd = out_req.fd;
        } else {
            self.out_lines_fd = -1;
        }

        if (in_lines.len > 0) {
            var in_req = std.mem.zeroes(c.gpio_v2_line_request);
            in_req.config.flags = c.GPIO_V2_LINE_FLAG_INPUT;
            in_req.config.num_attrs = 0; // do we need attributes?
            @memcpy(in_req.offsets[0..in_lines.len], in_lines);
            in_req.num_lines = @intCast(in_lines.len);
            @memcpy(in_req.consumer[0..gpio_consumer_name.len], gpio_consumer_name);

            _ = try misc.ioctl_checked(
                fd,
                c.GPIO_V2_GET_LINE_IOCTL,
                @intFromPtr(&in_req),
                "couldn't open GPIO input",
            );
            if (in_req.fd < 0) {
                std.log.err(
                    "Failed to open gpio device {s} input lines with errno = {d}",
                    .{ dev_path, std.c._errno().* },
                );
                return error.GpioLineOpenFailed;
            }
            self.in_lines_fd = in_req.fd;
        } else {
            self.in_lines_fd = -1;
        }

        return self;
    }

    pub fn deinit(self: *const GpioCtrl, io: std.Io) void {
        if (self.out_lines_fd >= 0) {
            _ = std.c.close(self.out_lines_fd);
        }
        if (self.in_lines_fd >= 0) {
            _ = std.c.close(self.in_lines_fd);
        }
        self.dev_file.close(io);
    }

    pub fn set(self: *GpioCtrl, val: u64, mask: u64) !void {
        var vals = std.mem.zeroes(c.gpio_v2_line_value);
        vals.bits = val;
        vals.mask = mask;
        _ = try misc.ioctl_checked(
            self.dev_file.handle,
            c.GPIO_V2_LINE_SET_VALUES_IOCTL,
            @intFromPtr(&vals),
            "couldn't set GPIO values",
        );
    }

    pub fn get(self: *GpioCtrl, mask: u64) !u64 {
        var vals = std.mem.zeroes(c.gpio_v2_line_value);
        vals.mask = mask;
        _ = try misc.ioctl_checked(
            self.dev_file.handle,
            c.GPIO_V2_LINE_GET_VALUES_IOCTL,
            @intFromPtr(&vals),
            "couldn't get GPIO values",
        );
        return vals.bits;
    }
};
